### PR TITLE
Fix non-ascii absolute file path handling in Uri

### LINKED ISSUE #1061: [Uri] Paths with Unicode/UTF-8 incorrectly parsed/reported by System.Uri
# Paths with Unicode/UTF-8 incorrectly parsed/reported by System.Uri

# General

As described by the title, paths with Unicode/UTF-8 characters are incorrectly parsed/reported by System.Uri resulting in an invalid path. For example the path "/üri" will result in an Uri like "file:///%C3%BCri/üri" (note the unescaped /üri at the end).

This also happens with other Unicode/UTF-8 characters like £, §, etc. So you can replace ü by any other Unicode/UTF-8 character in my example and see the same result, e.g. the path is doubled.

# Expected Result

```csharp
PathAndQuery = AbsolutePath = "/%C3%BCri"
AbsoluteUri = "file:///%C3%BCri"
```

Results collected using mono 5.14 on the same Linux machine.

# Actual Result

```csharp
PathAndQuery = AbsolutePath = "/%C3%BCri" // This seems to be correct
AbsoluteUri = "/%C3%BCri/%C3%BCri" // Note the additional "/%C3%BCri" at the end
_string = "/üri/üri" // Note the additinal "/üri" at the end
```

Using the .NET core version mentioned below.

# System Informations

```
$ dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   2.1.302
 Commit:    9048955601

Runtime Environment:
 OS Name:     gentoo
 OS Version:
 OS Platform: Linux
 RID:         gentoo-x64
 Base Path:   /opt/dotnet_core/sdk/2.1.302/

Host (useful for support):
  Version: 2.1.2
  Commit:  811c3ce6c0
```

# Code to reproduce

```csharp
using System;

namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            var uri = new Uri("/üri");
            Console.WriteLine(uri.ToString()); // file:///%C3%BCri/üri
        }
    }
}
```

Above code prints **"file:///%C3%BCri/üri"** while **"file:///%C3%BCri"** is expected.

Using **"/üri/üri"** in the Uri ctor results in a path like **"file:///%C3%BCri/%C3%BCri/üri/üri"**.

### PR DESCRIPTION
**Pull request by MihaZupan**
Fixes #1061

Actual fix is changing `IsDosPath` to `IsFile && !IsUncPath` when calculating path offsets.

### CODE REVIEW

File: src/libraries/System.Private.Uri/src/System/Uri.cs [hunk 1]
```diff
@@ -3262,16 +3262,17 @@ private unsafe void ParseRemaining()
             // For iri parsing if we found unicode the idx has offset into _originalUnicodeString..
             // so restart parsing from there and make _info.Offset.Path as _string.Length
 
-            idx = _info.Offset.Path;
-            origIdx = _info.Offset.Path;
+            idx = origIdx = _info.Offset.Path;
```
Comment (ManickaP) [inline src/libraries/System.Private.Uri/src/System/Uri.cs:3265]: Why this change?

File: src/libraries/System.Private.Uri/src/System/Uri.cs [hunk 2]
```diff
@@ -3262,16 +3262,17 @@ private unsafe void ParseRemaining()
...
 
             //Some uris do not have a query
             //    When '?' is passed as delimiter, then it's special case
             //    so both '?' and '#' will work as delimiters
             if (buildIriStringFromPath)
             {
-                // Dos paths have no host.  Other schemes cleared/set _string with host information in PrivateParseMinimal.
-                if (IsDosPath)
+                DebugAssertInCtor();
```
Comment (ManickaP) [inline src/libraries/System.Private.Uri/src/System/Uri.cs:3272]: So this method can also be called from outside of ctor but not in `buildIriStringFromPath` case?
Also, why is the assert repeated in every `if (buildIriStringFromPath)`?

File: src/libraries/System.Private.Uri/src/System/Uri.cs [hunk 3]
```diff
@@ -3314,19 +3315,7 @@ private unsafe void ParseRemaining()
...
-                try
-                {
-                    _string += escapedPath;
-                }
-                catch (ArgumentException)
-                {
-                    UriFormatException e = GetException(ParsingError.BadFormat)!;
-                    throw e;
-                }
+                _string += EscapeUnescapeIri(_originalUnicodeString, offset, origIdx, UriComponents.Path);
```
Comment (ManickaP) [inline src/libraries/System.Private.Uri/src/System/Uri.cs:3318]: The try-catch block removal is, I presume, due to `+=` never actually throwing?

**Review by ManickaP — COMMENTED** [review]
Just curious questions.

File: src/libraries/System.Private.Uri/src/System/Uri.cs (hunk 3 above)
Comment (MihaZupan) [reply src/libraries/System.Private.Uri/src/System/Uri.cs:3318]: Exactly, it's just dead code

File: src/libraries/System.Private.Uri/src/System/Uri.cs (hunk 2 above)
Comment (MihaZupan) [reply src/libraries/System.Private.Uri/src/System/Uri.cs:3272]: Correct. Uri lazily parses the input in 3 steps (ctor, CreateUriInfo, ParseRemaining). The latter two are done lazily as needed.

In case the Uri contains non-ascii characters, a full parse will be triggered from the constructor, so the `buildIriStringFromPath` branch can only run in the ctor. It also has to, as it updates the `_string` field in a non-synchronized manner.

The `DebugAssertInCtor` method was added recently (#33042) to ensure thread safety assumptions hold, I just missed this specific part.

> Also, why is the assert repeated in every if (buildIriStringFromPath)

It's not strictly needed, I just prefer it over a comment emphasizing the same thing.

File: src/libraries/System.Private.Uri/src/System/Uri.cs (hunk 1 above)
Comment (MihaZupan) [reply src/libraries/System.Private.Uri/src/System/Uri.cs:3265]: I find this way makes it more obvious that the same value is assigned to both variables.

I can revert this as it has no impact otherwise.

File: src/libraries/System.Private.Uri/src/System/Uri.cs (hunk 2 above)
Comment (ManickaP) [reply src/libraries/System.Private.Uri/src/System/Uri.cs:3273]: Thanks for explanation.
> It's not strictly needed, I just prefer it over a comment emphasizing the same thing.
Perfectly fine with me.

File: src/libraries/System.Private.Uri/tests/FunctionalTests/UriTests.cs [hunk 4]
```diff
@@ -738,5 +738,58 @@ public static void Uri_DoesNotLockOnString()
             Assert.True(Monitor.TryEnter(uriString, TimeSpan.FromSeconds(10)));
             Assert.False(timedOut);
         }
+
+        public static IEnumerable<object[]> FilePathHandlesNonAscii_TestData()
+        {
+            if (PlatformDetection.IsNotWindows)
```
Comment (ManickaP) [inline src/libraries/System.Private.Uri/tests/FunctionalTests/UriTests.cs:744]: Does this mean that `Uri` cannot handle Unix paths on Windows and vice-versa? Does the parsing itself depends on what OS the code is running?

**Review by ManickaP — APPROVED** [review]
One last question, otherwise LGTM.

File: src/libraries/System.Private.Uri/tests/FunctionalTests/UriTests.cs (hunk 4 above)
Comment (MihaZupan) [reply src/libraries/System.Private.Uri/tests/FunctionalTests/UriTests.cs:744]: > Does this mean that Uri cannot handle Unix paths on Windows and vice-versa?

Unix paths in implicit form will not be recognized on Windows.
`new Uri("/foo")` will throw on Windows and succeed on Unix.

DOS paths (`C:/foo`) are recognized the same on any platform.

> Does the parsing itself depend on what OS the code is running?

Unfortunately, yes.

Comment (ManickaP) [reply src/libraries/System.Private.Uri/tests/FunctionalTests/UriTests.cs:744]: Thanks for explanation!

**Review by wfurt — APPROVED** [review]
LGTM.

Should this be considered as breaking change? If so, we may add label to make it clear. 

Also test failures seems relevant.

**Comment by MihaZupan** [comment]
> Also test failures seems relevant.

Which ones? All I'm seeing rn is 1 cancelled, presumably from #34472

**Comment by MihaZupan** [comment]
> Should this be considered as breaking change?

Technically yes, but all the changed outputs go from outputting unusable gibberish to something reasonable.

**Comment by wfurt** [comment]
> Which ones?

It seems like I had stale links. With page reload veering looks good.

File: src/libraries/System.Private.Uri/tests/FunctionalTests/UriTests.cs [hunk 5]
```diff
@@ -738,5 +738,58 @@ public static void Uri_DoesNotLockOnString()
...
+            {
+                yield return new object[] { "file:///a/?b/c\u00FC/", "file:///a/?b/c\u00FC/", "/a/", "file:///a/?b/c%C3%BC/", "/a/" };
+                yield return new object[] { "file:///a/#b/c\u00FC/", "file:///a/#b/c\u00FC/", "/a/", "file:///a/#b/c%C3%BC/", "/a/" };
+                yield return new object[] { "file:///a/?b/#c/d\u00FC/", "file:///a/?b/#c/d\u00FC/", "/a/", "file:///a/?b/#c/d%C3%BC/", "/a/" };
+            }
+            else
+            {
+                yield return new object[] { "/a/?b/c\u00FC/", "file:///a/%3Fb/c\u00FC/", "/a/%3Fb/c%C3%BC/", "file:///a/%3Fb/c%C3%BC/", "/a/?b/c\u00FC/" };
+                yield return new object[] { "/a/#b/c\u00FC/", "file:///a/%23b/c\u00FC/", "/a/%23b/c%C3%BC/", "file:///a/%23b/c%C3%BC/", "/a/#b/c\u00FC/" };
+                yield return new object[] { "/a/?b/#c/d\u00FC/", "file:///a/%3Fb/%23c/d\u00FC/", "/a/%3Fb/%23c/d%C3%BC/", "file:///a/%3Fb/%23c/d%C3%BC/", "/a/?b/#c/d\u00FC/" };
```
Comment (lewing) [inline src/libraries/System.Private.Uri/tests/FunctionalTests/UriTests.cs:775]: Could you also add tests that start with the AbsoluteUri versions of these three?

Comment (MihaZupan) [reply src/libraries/System.Private.Uri/tests/FunctionalTests/UriTests.cs:775]: Can you show an example of what you mean? These are absolute Uris.

Comment (lewing) [reply src/libraries/System.Private.Uri/tests/FunctionalTests/UriTests.cs:775]: I mean checking that a well formed string file: url can resolve back to a local file by checking the results of:
```
                yield return new object[] {"file:///a/%3Fb/c%C3%BC/" ...
                yield return new object[] {"file:///a/%23b/c%C3%BC/"...
                yield return new object[] {""file:///a/%3Fb/%23c/d%C3%BC/"...
```

Comment (lewing) [reply src/libraries/System.Private.Uri/tests/FunctionalTests/UriTests.cs:775]: even better make FilePathHandlesNonAscii check that new Uri (uri.AbsoluteUri) and Uri (uri.LocalPath) pass all the same checks that Uri (uriString) does

**Comment by davidsh** [comment]
>Technically yes, but all the changed outputs go from outputting unusable gibberish to something reasonable.

It still should be marked as a breaking change since we are changing the output. We might consider it an "acceptable" breaking change. But many customers haven't got broken when we "fix" System.Uri. So, we should document it accordingly.

File: src/libraries/System.Private.Uri/tests/FunctionalTests/UriTests.cs (hunk 5 above)
Comment (MihaZupan) [reply src/libraries/System.Private.Uri/tests/FunctionalTests/UriTests.cs:775]: I added the extra assertions around `new Uri (uri.AbsoluteUri)`. `LocalPath`, however, can not roundtrip as it drops information in some cases.

For example `file:///a/?b/c\u00FC/`, `LocalPath:` `/a/`

**Comment by danmosemsft** [comment]
@MihaZupan this is labeled breaking change. Can you please make sure we have an issue with https://github.com/dotnet/docs/issues/new?template=dotnet-breaking-change.md so that it appears in https://github.com/dotnet/docs/issues?q=is%3Aopen+is%3Aissue+label%3Abreaking-change ? I don't see it.
